fetch/cosmwasm_contract/src/contract.rs

```diff
@@ -1,69 +1,398 @@
+        HandleMsg::Swap { destination } => {
+            let amount = amount_from_funds(&env.message.sent_funds);
+            try_swap(deps, &env, &state, amount, destination)
+        }
+        HandleMsg::ReverseSwap {
+            rid,
+            to,
+            from,
+            origin_tx_hash,
+            amount,
+            relay_eon,
+        } => try_reverse_swap(
+            deps,
+            &env,
+            &state,
+            rid,
+            to,
+            from,
+            origin_tx_hash,
+            amount,
+            relay_eon,
+        ),
+        HandleMsg::Refund {
+            id: _,
+            to: _,
+            amount: _,
+            relay_eon: _,
+        } => Ok(HandleResponse::default()),
         HandleMsg::Pause { since_block: _ } => Ok(HandleResponse::default()),
-        HandleMsg::FreezeFunds {amount: _}  => Ok(HandleResponse::default()),
-        HandleMsg::UnFreezeFunds {amount: _}  => Ok(HandleResponse::default()),
+        HandleMsg::FreezeFunds { amount: _ } => Ok(HandleResponse::default()),
+        HandleMsg::UnFreezeFunds { amount: _ } => Ok(HandleResponse::default()),
         HandleMsg::SetCap { amount: _ } => Ok(HandleResponse::default()),
-        HandleMsg::SetLimits { swap_min: _, swap_max: _, swap_fee: _ } => Ok(HandleResponse::default()),
-        HandleMsg::GrantRole { role: _, address: _ } => Ok(HandleResponse::default()),
-        HandleMsg::RevokeRole { role: _, address: _ } => Ok(HandleResponse::default()),
-        HandleMsg::RenounceRole { role: _, address: _ } => Ok(HandleResponse::default()),
+        HandleMsg::SetLimits {
+            swap_min: _,
+            swap_max: _,
+            swap_fee: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::GrantRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::RevokeRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::RenounceRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
     }
 }
 
 fn try_swap<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
-    _env: Env,
-    _amount: U256,
-    _destination: String,
+    env: &Env,
+    state: &State,
+    amount: Uint128,
+    destination: String,
 ) -> StdResult<HandleResponse> {
-   config(&mut deps.storage).update(|mut state| {
+    verify_not_paused(env, state)?;
+    verify_swap_amount(amount, state)?;
+
+    let swap_id = state.next_swap_id;
+    config(&mut deps.storage).update(|mut state| {
+        state.supply += amount;
         state.next_swap_id += 1;
         Ok(state)
     })?;
 
-    Ok(HandleResponse::default())
+    let log = vec![
+        log("action", "swap"),
+        log("destination", destination),
+        log("swap_id", swap_id),
+        log("amount", amount), // TOFIX(LR) how about fees?
```

Comment:
Fees will be ultimatelly resoled on the other blockchain in counterpart contract, but here in this contracty we have the `verify_swap_amount(...)` check, which ensures that `amount >= lower_swap_limit`, where it is implicitly assumed that we set `lower_swap_limit` value in **this** contrac so that it will be equal or higher than `swapFee` value in **countyerpart** conract on the **OTHER** blockchain.